From d51dd68c567222ca383e5c64a9ad7f45b6e0832a Mon Sep 17 00:00:00 2001
From: Zheng Zucheng <zhengzucheng@huawei.com>
Date: Thu, 25 May 2023 15:43:58 +0800
Subject: [PATCH 236/780] [Huawei] sched: Add dynamic switch for normalize util
 feature

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188931

--------------------------------

Add /proc/sys/kernel/sched_normalize_util to enable normalize util,
0 to disable cpu normalize util, 1 to enable cpu normalize util, default
value is 0.

Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 include/linux/sched/sysctl.h |  1 +
 kernel/sched/core.c          |  7 +++++++
 kernel/sched/cpuacct.c       |  3 +++
 kernel/sysctl.c              | 23 +++++++++++++++++++++++
 4 files changed, 34 insertions(+)

diff --git a/include/linux/sched/sysctl.h b/include/linux/sched/sysctl.h
index a829bbf5ddd1..e59b2e7400a6 100644
--- a/include/linux/sched/sysctl.h
+++ b/include/linux/sched/sysctl.h
@@ -38,6 +38,7 @@ extern int sysctl_sched_util_low_pct;
 #ifdef CONFIG_QOS_NORMALIZE_UTIL
 extern unsigned int sysctl_multi_task_factor;
 extern unsigned int sysctl_single_task_factor;
+DECLARE_STATIC_KEY_FALSE(sysctl_normalize_util_key);
 #endif
 
 enum sched_tunable_scaling {
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 658fe7ad45bc..bf631c8fcca0 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -4098,6 +4098,7 @@ void scheduler_tick(void)
 }
 
 #ifdef CONFIG_QOS_NORMALIZE_UTIL
+DEFINE_STATIC_KEY_FALSE(sysctl_normalize_util_key);
 DEFINE_PER_CPU(struct smt_peer_stat, smt_util_stat);
 void __init init_normalize_util(int cpu)
 {
@@ -4115,6 +4116,9 @@ static void update_local_normalize_runtime(struct task_struct *tsk, int cpu)
 	struct rq *rq = NULL;
 	struct smt_peer_stat *stat = NULL;
 
+	if (!static_branch_unlikely(&sysctl_normalize_util_key))
+		return;
+
 	rq = cpu_rq(cpu);
 	stat = &per_cpu(smt_util_stat, cpu);
 	raw_spin_lock(&stat->lock);
@@ -4142,6 +4146,9 @@ void update_sibling_normalize_runtime(bool enter)
 	struct rq *sibling_rq = NULL;
 	struct smt_peer_stat *pstat = NULL;
 
+	if (!static_branch_unlikely(&sysctl_normalize_util_key))
+		return;
+
 	sibling = this_cpu_read(cpu_sibling);
 	if (unlikely(!sched_smt_active() || sibling == -1 ||
 		     cpu_is_offline(sibling)))
diff --git a/kernel/sched/cpuacct.c b/kernel/sched/cpuacct.c
index a62ccff5ab14..bbc84f09c497 100644
--- a/kernel/sched/cpuacct.c
+++ b/kernel/sched/cpuacct.c
@@ -371,6 +371,9 @@ void cpuacct_normalize_charge(struct task_struct *tsk, u64 cputime)
 	u64 ptime;
 	struct cpuacct *ca;
 
+	if (!static_branch_unlikely(&sysctl_normalize_util_key))
+		return;
+
 	ptime = cpuusage_sibling_runtime(task_cpu(tsk));
 	if (unlikely(ptime > cputime))
 		ptime = cputime;
diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index bbbb9a7bc869..6a3b98210ffe 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -1752,6 +1752,22 @@ int proc_do_static_key(struct ctl_table *table, int write,
 	return ret;
 }
 
+#ifdef CONFIG_QOS_NORMALIZE_UTIL
+int proc_do_normalize_util_switch(struct ctl_table *table, int write,
+				  void *buffer, size_t *lenp, loff_t *ppos)
+{
+	int cpu;
+
+	for_each_online_cpu(cpu) {
+		if (cpumask_weight(cpu_smt_mask(cpu)) > 2) {
+			pr_warn_once("Not support SMT normalize util\n");
+			return -EINVAL;
+		}
+	}
+
+	return proc_do_static_key(table, write, buffer, lenp, ppos);
+}
+#endif
 #ifdef CONFIG_IAS
 static struct ctl_table ias_table[] = {
 #ifdef CONFIG_IAS_SMART_IDLE
@@ -2877,6 +2893,13 @@ static struct ctl_table kern_table[] = {
 		.extra1         = SYSCTL_ONE,
 		.extra2         = &one_hundred,
 	},
+	{
+		.procname	= "sched_normalize_util",
+		.data		= &sysctl_normalize_util_key.key,
+		.maxlen         = sizeof(sysctl_normalize_util_key),
+		.mode		= 0644,
+		.proc_handler	= proc_do_normalize_util_switch,
+	},
 #endif
 	{ }
 };
-- 
2.17.1

